### PR TITLE
[MPS] Ignore bias in complex `[b]add[b]mm` when beta is 0

### LINKED ISSUE #194442: [MPS] torch.baddbmm with beta=0 propagates NaN from input, violating documented semantics
### 🐛 Describe the bug

`torch.baddbmm` on MPS propagates NaN from `input` even when `beta=0`. The documentation states:

> If β is 0, then the content of `input` will be ignored, and nan and infinity in it will not be propagated to the output.

On MPS this contract is violated: a single NaN scalar in `input` (shape `(1,1,1)` broadcast) turns the **entire** output into NaN.

### Reproduction

```python
import torch

q = torch.randn(1, 64, 128, device="mps", dtype=torch.bfloat16)
k = torch.randn(1, 128, 3520, device="mps", dtype=torch.bfloat16)

nan_in = torch.full((1, 1, 1), float("nan"), device="mps", dtype=torch.bfloat16)
out = torch.baddbmm(nan_in, q, k, alpha=0.088, beta=0)
print(int(torch.isnan(out).sum()), "/", out.numel())
```

Observed output:

```
225280 / 225280        # MPS: every element is NaN
```

Same code on CPU prints `0`, matching the documented semantics.

Note: with real-world usage the NaN input is not deliberately constructed — it comes from *uninitialized* memory. Any code that initializes the accumulator via `torch.empty(...)` and relies on `beta=0` to ignore it will intermittently read garbage bit patterns that decode to NaN/Inf.

### Real-world impact

ComfyUI's sub-quadratic attention (`comfy/ldm/modules/sub_quadratic_attention.py`) computes attention scores exactly this way:

```python
attn_weights = torch.baddbmm(
    torch.empty(1, 1, 1, device=query.device, dtype=query.dtype),
    query, key_t, alpha=scale, beta=0,
)
```

On Apple Silicon this intermittently poisons whole attention score matrices whenever the recycled memory behind `empty(1,1,1)` contains NaN patterns → diffusion sampling collapses → video models produce all-black outputs at very high rates (~73% of runs observed with LTX-2.x bf16 workflows; deterministic per-seed because allocation layout is deterministic). Initializing with `zeros()` instead of `empty()` eliminates the failures completely.

### Environment

- PyTorch version: 2.13.0
- OS: macOS 26.6.2 (Darwin 25.6.0), arm64
- Device: Apple M5 Max (MPS)
- Python: 3.13.12

Any CPU/CUDA backend honors the documented semantics; MPS appears to multiply by `beta` without special-casing `beta == 0`.

cc @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

### PR DESCRIPTION
**Pull request by Achraf921**
Fixes #194442.

`baddbmm`/`addbmm`/`addmm` [document](https://docs.pytorch.org/docs/stable/generated/torch.addmm.html) that when `beta == 0` the `input`/bias is ignored, so nan/inf in it must not propagate. On MPS, real dtypes already honor this (the MPSGraph path drops the bias node when `beta == 0`, and the bfloat16 case from the issue was already fixed on main after 2.13.0), but complex and integral types take the Metal-kernel path, which unconditionally adds `beta*bias`, so a nan/inf `input` poisons the whole output even though `beta == 0`.

When `beta == 0` the bias must be dropped entirely:
- baddbmm lowers to a plain `bmm` and addmm to a plain `mm`, scaling in place only for the uncommon `alpha != 1`
- addbmm has no bias-free "batched matmul + reduce" kernel, so it stays on its own kernel but is handed a broadcast scalar-zero bias

It also fixes complex `addmm` more broadly: the MPSGraph addmm path applied `alpha`/`beta` via `toDouble()`, which throws for a complex scalar and silently drops the imaginary part (and a pure-imaginary `beta` was mis-detected as zero, wrongly dropping the bias). Complex `addmm` is now routed through the Metal kernel like `bmm`/`baddbmm`. Integral types cannot carry nan, so nothing changes for them.

For complex baddbmm with `beta=0` and a nan `input`, the poisoned-output count goes from `256 / 256` on main to `0 / 256`, matching CPU; same for addbmm/addmm. Complex `alpha`/`beta` addmm now matches CPU instead of raising.

Authored with the assistance of an AI agent (Claude Code).

Co-authored-by: Achraf Bayi <achrafbayi@icloud.com>

### CODE REVIEW

**Review by malfet — CHANGES_REQUESTED** [review]
Can you explain what's the point of MPSStream changes?
Also test?

**Comment by Achraf921** [comment]
> Can you explain what's the point of MPSStream changes? Also test?

MPSStream changes were a branching mistake on my side, removed them, also extended the regression test to complex64 values, which fails on the pre-fix build

File: aten/src/ATen/native/mps/operations/LinearAlgebra.mm [hunk 1]
```diff
@@ -1200,7 +1200,16 @@ static void linalg_inv_ex_out_mps_impl(const Tensor& A, bool check_errors, const
 
   // Use Metal kernels for integer and complex types
   if (c10::isIntegralType(batch1.scalar_type(), true) || c10::isComplexType(batch1.scalar_type())) {
-    return do_metal_addbmm_or_baddbmm(input, batch1, batch2, alpha, beta, result, opType == BADDBMM_OP_TYPE);
+    // beta == 0 means input is ignored, so nan/inf in it must not propagate (see the
+    // MPSGraph path below, which drops the bias node entirely). The Metal kernel always
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/LinearAlgebra.mm]: Which MPSGraph  path are you talking about?

**Comment by malfet** [comment]
> Tested on an M4, macOS 26.6.1. Full test/test_mps.py shows the same 3
failures and 15 errors with and without the change, all pre-existing on this
base.

Do you mind creating an issue reporting those? As it should not :)

File: aten/src/ATen/native/mps/operations/LinearAlgebra.mm [hunk 2]
```diff
@@ -1200,7 +1200,16 @@ static void linalg_inv_ex_out_mps_impl(const Tensor& A, bool check_errors, const
...
+    // reads the bias, so hand it zeros instead.
+    return do_metal_addbmm_or_baddbmm((beta.toComplexDouble() == 0.0 ? (at::zeros_like(input)) : input),
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/LinearAlgebra.mm]: This is expensive and wasteful. Much better option would be to redispatch to BMM_OP_TYPE if alpha is also 1, which would be most common, or if you need to create an single element tensor which will be broadcast, rather than copy the entire input

**Review by malfet — CHANGES_REQUESTED** [review]
Looks much better, but your PR body comment no longer aligns with the actual fix. And also, fix could have been made slightly better, by redispatching to add_mm and calling in place mul in rare case when alpha is not 1.0

**Review by malfet — APPROVED** [review]
